### PR TITLE
[Merged by Bors] - feat(Order/Cofinal): predicate for cofinal sets

### PR DESCRIPTION
**Pull request by vihdzp**
The plan is of course to redefine `Order.cof` as the least cardinality of a `Cofinal` set.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by YaelDillies — COMMENTED** [review]
What's your plan with respect to `Order.Cofinal`? Do you know it exists?

**Comment by vihdzp** [comment]
Oh, I didn't know that existed. I'm not really sure what that entails. Would it make sense to have a bundled and an unbundled version?

**Comment by YaelDillies** [comment]
Since cofinal sets have so few properties (you can't take their intersection for example), I would think we don't need the type.

**Comment by vihdzp** [comment]
Yeah, I realize I could define `Order.cof α = ⨆ s : Order.Cofinal α, #s`.

Still, I would like to have some sort of API - `Order.Cofinal` provides little to work with. I'll ask on Zulip to see what other people think.

**Comment by vihdzp** [comment]
https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/Cofinal.20sets/near/478869434

File: Mathlib/Order/Cofinal.lean [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+Copyright (c) 2024 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.Order.Bounds.Basic
+
+/-!
+# Cofinal sets
+
+A set `s` is cofinal when for every `a` there exists an element of `s` greater or equal to it. This
```
Comment (AntoineChambert-Loir) [inline Mathlib/Order/Cofinal.lean:11]: A set in what kind of type? Maybe add `Let alpha be a type with [LE alpha]`.

Comment (YaelDillies) [reply Mathlib/Order/Cofinal.lean:11]: I don't think you should be bombarding your reader with Lean notation from the first sentence of the docstring. What about
```suggestion
A set `s` in an order is cofinal when for every `a` there exists an element of `s` greater or equal to it. This
```

Comment (AntoineChambert-Loir) [reply Mathlib/Order/Cofinal.lean:11]: “in an ordered type” seems fine to me. At least the reader has to be told what are the implicit datas used in the definition, and not to guess depending on the future syntax. 
(And “bombarded” is of the passive-aggressive kind of vocabulary that should be avoided in written discussion.)

Comment (vihdzp) [reply Mathlib/Order/Cofinal.lean]: I've rewritten this using the "ordered type" wording.

**Comment by vihdzp** [comment]
> What's your plan with respect to `Order.Cofinal`? Do you know it exists? @YaelDillies 

I think the conclusion of the Zulip discussion is that `Order.Cofinal` should be refactored in favor of this predicate. That can happen in a follow-up PR.

File: Mathlib/Order/Cofinal.lean [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+- Define `Order.cof` in terms of `Cofinal`.
+- Deprecate `Order.Cofinal` in favor of this predicate.
+-/
+
+variable {α : Type*}
+
+section LE
+variable [LE α]
+
+theorem isCofinal_of_isEmpty [IsEmpty α] (s : Set α) : IsCofinal s :=
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:29]:
```suggestion
theorem IsCofinal.of_isEmpty [IsEmpty α] (s : Set α) : IsCofinal s :=
```

File: Mathlib/Order/Cofinal.lean [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  fun a ↦ isEmptyElim a
+
+theorem isCofinal_empty_iff : IsCofinal (∅ : Set α) ↔ IsEmpty α := by
+  refine ⟨fun h ↦ ⟨fun a ↦ ?_⟩, fun h ↦ isCofinal_of_isEmpty _⟩
+  simpa using h a
+
+theorem isCofinal_top [OrderTop α] : IsCofinal {(⊤ : α)} :=
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:36]:
```suggestion
theorem IsCofinal.singleton_top [OrderTop α] : IsCofinal {(⊤ : α)} :=
```

File: Mathlib/Order/Cofinal.lean [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  fun _ ↦ ⟨⊤, Set.mem_singleton _, le_top⟩
+
+theorem isCofinal_mono {s t : Set α} (h : s ⊆ t) (hs : IsCofinal s) : IsCofinal t := by
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:39]:
```suggestion
theorem IsCofinal.mono {s t : Set α} (h : s ⊆ t) (hs : IsCofinal s) : IsCofinal t := by
```

File: Mathlib/Order/Cofinal.lean [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  intro a
+  obtain ⟨b, hb, hb'⟩ := hs a
+  exact ⟨b, h hb, hb'⟩
+
+end LE
+
+section Preorder
+variable [Preorder α]
+
+theorem isCofinal_univ : IsCofinal (@Set.univ α) :=
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:49]:
```suggestion
theorem IsCofinal.univ : IsCofinal (@Set.univ α) :=
```

File: Mathlib/Order/Cofinal.lean [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  obtain ⟨b, hb, hb'⟩ := hs a
+  obtain ⟨c, hc, hc'⟩ := ht ⟨b, hb⟩
+  exact ⟨c, Set.mem_image_of_mem _ hc, hb'.trans hc'⟩
+
+end Preorder
+
+section PartialOrder
+variable [PartialOrder α]
+
+theorem IsCofinal.isMax_mem {s : Set α} {a : α} (ha : IsMax a) (hs : IsCofinal s) : a ∈ s := by
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:68]:
```suggestion
theorem IsCofinal.mem_of_isMax {s : Set α} {a : α} (ha : IsMax a) (hs : IsCofinal s) : a ∈ s := by
```
maybe?

File: Mathlib/Order/Cofinal.lean [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  obtain ⟨b, hb, hb'⟩ := hs a
+  rwa [ha.eq_of_ge hb'] at hb
+
+theorem IsCofinal.top_mem [OrderTop α] {s : Set α} (hs : IsCofinal s) : ⊤ ∈ s :=
+  hs.isMax_mem isMax_top
+
+@[simp]
+theorem isCofinal_iff_of_orderTop [OrderTop α] {s : Set α} : IsCofinal s ↔ ⊤ ∈ s :=
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:76]:
```suggestion
theorem isCofinal_iff_top_mem [OrderTop α] {s : Set α} : IsCofinal s ↔ ⊤ ∈ s :=
```

**Review by YaelDillies — APPROVED** [review]
Okay then

maintainer delegate

Thanks!

File: Mathlib/Order/Cofinal.lean [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+end PartialOrder
+
+section LinearOrder
+variable [LinearOrder α]
+
+theorem not_isCofinal_iff {s : Set α} : ¬ IsCofinal s ↔ ∃ x, ∀ y ∈ s, y < x := by
+  simp [IsCofinal]
+
+theorem bddAbove_of_not_isCofinal {s : Set α} (h : ¬ IsCofinal s) : BddAbove s := by
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:87]:
```suggestion
theorem BddAbove.of_not_isCofinal {s : Set α} (h : ¬ IsCofinal s) : BddAbove s := by
```

File: Mathlib/Order/Cofinal.lean [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  rw [not_isCofinal_iff] at h
+  obtain ⟨x, h⟩ := h
+  exact ⟨x, fun y hy ↦ (h y hy).le⟩
+
+theorem isCofinal_of_not_bddAbove {s : Set α} (h : ¬ BddAbove s) : IsCofinal s := by
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:92]:
```suggestion
theorem IsCofinal.of_not_bddAbove {s : Set α} (h : ¬ BddAbove s) : IsCofinal s := by
```

File: Mathlib/Order/Cofinal.lean [hunk 10]
```diff
@@ -30,7 +30,7 @@ theorem IsCofinal.of_isEmpty [IsEmpty α] (s : Set α) : IsCofinal s :=
   fun a ↦ isEmptyElim a
 
 theorem isCofinal_empty_iff : IsCofinal (∅ : Set α) ↔ IsEmpty α := by
-  refine ⟨fun h ↦ ⟨fun a ↦ ?_⟩, fun h ↦ isCofinal_of_isEmpty _⟩
+  refine ⟨fun h ↦ ⟨fun a ↦ ?_⟩, fun h ↦ IsCofinal.of_isEmpty _⟩
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:33]:
```suggestion
  refine ⟨fun h ↦ ⟨fun a ↦ ?_⟩, fun h ↦ .of_isEmpty _⟩
```

File: Mathlib/Order/Cofinal.lean [hunk 11]
```diff
@@ -50,7 +50,7 @@ theorem IsCofinal.univ : IsCofinal (@Set.univ α) :=
   fun a ↦ ⟨a, ⟨⟩, le_rfl⟩
 
 instance : Inhabited {s : Set α // IsCofinal s} :=
-  ⟨_, isCofinal_univ⟩
+  ⟨_, IsCofinal.univ⟩
```
Comment (YaelDillies) [inline Mathlib/Order/Cofinal.lean:53]:
```suggestion
  ⟨_, .univ⟩
```
etc

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**